core/src/main/java/org/apache/cxf/common/spi/ClassGeneratorClassLoader.java

```diff
@@ -0,0 +1,151 @@
+            e.printStackTrace();
+        }
+    }
+    protected Class<?> loadClass(String className, byte[] bytes) {
+        if (DEBUG) {
+            saveClass(className, bytes);
+        }
+        TypeHelperClassLoader loader = getOrCreateLoader();
+        synchronized (loader) {
+            Class<?> cls = loader.lookupDefinedClass(className);
+            if (cls == null) {
+                return loader.defineClass(className, bytes);
+            }
+            return cls;
+        }
+    }
+    protected Class<?> findClass(String className) {
+        return getOrCreateLoader().lookupDefinedClass(className);
+    }
+    private TypeHelperClassLoader getOrCreateLoader() {
+        TypeHelperClassLoader loader = bus.getExtension(TypeHelperClassLoader.class);
+        if (loader == null) {
+            synchronized (bus) {
+                loader = bus.getExtension(TypeHelperClassLoader.class);
+                if (loader == null) {
+                    ClassLoader parent = bus.getExtension(ClassLoader.class);
+                    if (parent == null) {
+                        parent = Thread.currentThread().getContextClassLoader();
+                    }
+                    loader = new TypeHelperClassLoader(parent);
+                    bus.setExtension(loader, TypeHelperClassLoader.class);
+                }
+            }
+        }
+        return loader;
+    }
+
+
+    public static class TypeHelperClassLoader extends ClassLoader {
+        ConcurrentHashMap<String, Class<?>> defined = new ConcurrentHashMap<>();
+
+        TypeHelperClassLoader(ClassLoader parent) {
+            super(parent);
+        }
+        public Class<?> lookupDefinedClass(String name) {
+            return defined.get(name.replace('/', '.'));
+        }
+
+        @Override
+        protected Class<?> findClass(String name) throws ClassNotFoundException {
+            if (name.endsWith("package-info")) {
+                return getParent().loadClass(name);
+            }
+            return super.findClass(name);
+        }
+
+        public Class<?> defineClass(String name, byte[] bytes) {
+            Class<?> ret = defined.get(name.replace('/', '.'));
+            if (ret != null) {
+                return ret;
+            }
+            if (name.endsWith("package-info")) {
+                Package p = super.getPackage(name.substring(0, name.length() - 13));
+                if (p == null) {
+                    definePackage(name.substring(0, name.length() - 13).replace('/', '.'),
+                            null,
+                            null,
+                            null,
+                            null,
+                            null,
+                            null,
+                            null);
+                }
+            }
+
+            ret = super.defineClass(name.replace('/', '.'), bytes, 0, bytes.length);
+            Class<?> tmpRet = defined.putIfAbsent(name.replace('/', '.'), ret);
```

Comment:
Minor, probably `computeIfAbsent` would be a bit more helpful (may eliminate some calls to `super.defineClass`)

Reply:
done